core/src/test/java/org/elasticsearch/index/query/TermsQueryBuilderTest.java

```diff
@@ -0,0 +1,136 @@
+ */
+
+package org.elasticsearch.index.query;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.lucene.search.Queries;
+import org.elasticsearch.indices.cache.query.terms.TermsLookup;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.hamcrest.Matchers.is;
+
+public class TermsQueryBuilderTest extends BaseQueryTestCase<TermsQueryBuilder> {
+
+    @Override
+    protected TermsQueryBuilder doCreateTestQueryBuilder() {
+        TermsQueryBuilder query = newRandomQueryBuilder();
+
+        if (randomBoolean()) {
+            query.isFilter(true);
+        }
+
+        // NO COMMIT: also handle the terms lookup query case
+        return query;
+    }
+
+    private TermsQueryBuilder newRandomQueryBuilder() {
+        String fieldName = null;
+        List<Object> values = new ArrayList<>();
+        switch (randomIntBetween(0, 2)) {
+            case 0:
+                if (randomBoolean()) {
+                    fieldName = STRING_FIELD_NAME;
+                }
+                if (frequently()) {
+                    for (int i = 0; i < randomInt(5); i++) {
+                        values.add(randomAsciiOfLengthBetween(1, 10));
+                    }
+                } else {
+                    // generate unicode strings in 10% of cases
+                    for (int i = 0; i < randomInt(5); i++) {
+                        values.add(randomUnicodeOfLength(10));
+                    }
+                }
+                break;
+            case 1:
+                if (randomBoolean()) {
+                    fieldName = INT_FIELD_NAME;
+                }
+                for (int i = 0; i < randomInt(5); i++) {
+                    values.add(randomInt(10000));
+                }
+                break;
+            case 2:
+                if (randomBoolean()) {
+                    fieldName = DOUBLE_FIELD_NAME;
+                }
+                for (int i = 0; i < randomInt(5); i++) {
+                    values.add(randomDouble());
+                }
+                break;
+            default:
+                throw new UnsupportedOperationException();
+        }
+        if (fieldName == null) {
+            fieldName = randomAsciiOfLengthBetween(1, 10);
+        }
+        return new TermsQueryBuilder(fieldName, values);
+    }
+
+    @Override
+    protected Query doCreateExpectedQuery(TermsQueryBuilder queryBuilder, QueryParseContext context) throws IOException {
+        List<Object> terms = new ArrayList<>();
+        if (queryBuilder.isTermsLookupQuery()) {
+            TermsLookup termsLookup = new TermsLookup(queryBuilder.lookupIndex(), queryBuilder.lookupType(), queryBuilder.lookupId(), 
+                    queryBuilder.lookupRouting(), queryBuilder.lookupPath(), context);
+            // NO COMMIT: we would need to mock a client in BaseQueryTestCase here
+            // we could also more simply mock handleTermsLookup
```

Comment:
+2 for mocking, not sure how easy it is to mock a client, maybe mocking handleTermsLookup would work better?